### ISSUE TITLE
Ability to use a local copy of npm package

### ISSUE THREAD

**Issue by brc-dd** [denoland/deno#15624]
Is there any way to use a local (modified) copy of an NPM package instead of a published one in this command? Thanks!

```sh
deno run -A --unstable npm:vite
```

**Comment by dsherret** [comment]
No, it's not something that's possible at the moment, but definitely on our radar. It wouldn't be difficult to implement. It's more a question of how it's specified so if anyone has any ideas let us know.

Maybe like `npm:///path/to/local/package/folder`, then if you do `npm:///path/to/local/package/folder/some/sub_path.js` it would resolve the first actual package.json going up the parents (so at `/path/to/local/package/folder/package.json`) and then resolve the rest of the path (`./some/sub_path.js`) as a subpath from there. I'm not sure though and there's probably some way that breaks down.

**Comment by brc-dd** [comment]
For now, I suppose just some way to use tarball would be sufficient, one can just do npm pack and deno run. 🤔

**Comment by yisibl** [comment]
Yes, we need the ability to import local packages.

Usage scenario: I am an author of an npm package and I need to test if my npm package works well in deno.

**Comment by JonasBa** [comment]
+1 here, I am attempting to make test an npm package using native modules in deno and this would have made my life much easier

**Comment by JonasBa** [comment]
@dsherret mind providing some extra context around this, I might try and find some time to give this a go.

**Comment by dsherret** [comment]
This one probably requires about 1-2 weeks of work because we'd have to support having multiple paths for what we consider to be "node code" and it introduces a new concept in npm resolution where we need to get package information from the file system instead of just the npm registry and then also make this work with node resolution. I wouldn't recommend it as a first task.

In the meantime, for testing npm packages locally you can use the `--node-modules-dir` flag to cache an existing version of a package locally and make modifications to it. For example:

```
> cat main.ts
import CodeBlockWriter from "npm:code-block-writer@12.0.0";
console.log(CodeBlockWriter);
> deno cache --node-modules-dir main.ts
> # overwrite the contents of node_modules/.deno/code-block-writer@12.0.0/node_modules/code-block-writer here and make sure to leave the `.initialized` file
> deno run --node-modules-dir main.ts # this will now use your modifications
```

**Comment by brc-dd** [comment]
Something like verdaccio should also work along with `NPM_CONFIG_REGISTRY` env var 👀

**Comment by jedwards1211** [comment]
This would solve the problems with CI testing that npm packages work with Deno: https://github.com/denoland/deno/issues/19621

**Comment by alexgleason** [comment]
> In the meantime, for testing npm packages locally you can use the `--node-modules-dir` flag to cache an existing version of a package locally and make modifications to it. For example:

I could not get this to work at all until I added a `package.json` file containing an empty object `{}` in the same directory. Then it worked.

**Comment by mattt** [comment]
👋 I maintain the [replicate](https://github.com/replicate/replicate-javascript) package, and recently added integration tests for Deno, following @dsherret's suggestion in https://github.com/denoland/deno/issues/15624#issuecomment-1511621310 and @alexgleason's tip in https://github.com/denoland/deno/issues/15624#issuecomment-1967080476. 

Here's what worked for us:

- Create a tarball of the package using `npm --loglevel error pack` [^1]
- Create a Deno test project with the following files:

```jsonc
// deno.json
{
  "imports": {
    "replicate": "npm:replicate"
  }
}
```

```jsonc
// package.json
{}
```

The app runs a [hello-world](https://replicate.com/replicate/hello-world) model on Replicate that echoes back the string input sent to it. The test runs the app and tests its expected output.

```typescript
// index.ts
import Replicate from "replicate";

const replicate = new Replicate({
  auth: Deno.env.get("REPLICATE_API_TOKEN"),
});

export default async function main() {
  return await replicate.run(
    "replicate/hello-world:5c7d5dc6dd8bf75c1acaa8565735e7986bc5b66206b55cca93cb72c9bf15ccaa",
    {
      input: {
        text: "Deno the dinosaur",
      },
    }
  );
}
```

```type
// index.test.ts
import { assertEquals } from "jsr:@std/assert";
import main from "./index.ts";

Deno.test({
  name: "main",
  async fn() {
    const output = await main();
    assertEquals({ output }, { output: "hello Deno the dinosaur" });
  },
});
```

- In the Deno test project, run `deno cache --node-modules-dir index.ts` to initialize a `node_modules` directory
- Overwrite the `replicate` module with the contents of the tarball from before by running `tar -xzf path/to/tarball --strip-components=1 -C integration/deno/replicate` [^2]
- Invoke the test with `deno test --allow-env --allow-net --node-modules-dir index.test.ts`

It'd be nice for Deno to support [overriding npm modules](https://docs.deno.com/runtime/manual/basics/import_maps/#overriding-imports) directly from a tarball file. But for now, this seems to work for us, and I'm happy to have found a way to verify that our package supports Deno.

[^1]: In our [GitHub Actions workflow](https://github.com/replicate/replicate-javascript/blob/90de8f866f2f476a86e4585a18e18bebd3ce4a2a/.github/workflows/ci.yml) we have a [`build` job](https://github.com/replicate/replicate-javascript/blob/90de8f866f2f476a86e4585a18e18bebd3ce4a2a/.github/workflows/ci.yml#L32-L52) that uses `actions/upload-artifact` to use this tarball in integration tests
[^2]: `npm pack` creates an archive with the contents nested under a top-level `package` directory. The `--strip-components=1` option tells `tar` to write the contents of that directory into `node_modules/replicate` without nesting.

**Comment by jedwards1211** [comment]
@mattt I wonder if that works if the package has dependencies though?  I don't see any dependencies in `replicate`'s `package.json`.

**Comment by mattt** [comment]
@jedwards1211 Good question. I think it might. The `replicate` package has an [optional dependency on `readable-stream`](https://github.com/replicate/replicate-javascript/blob/4c04a57dcaab4ed1d74673ed74bf3ef33ab5b90c/package.json#L37C6-L37C21), and after applying this workflow, the `node_modules/.deno` directory includes a cache of `readable-stream@4.5.2`. So Deno is resolving that at least.

**Comment by ara4n** [comment]
This is stumping me (at least on deno 1.45.5 on macOS).  If I follow @dsherret's workaround in https://github.com/denoland/deno/issues/15624#issuecomment-1511621310 and symlink the equivalent of `node_modules/.deno/code-block-writer@12.0.0/node_modules/code-block-writer` to my local dev checkout at `~/workspace/code-block-writer`, it fails with:

`error: Loading unprepared module: file:///Users/matthew/workspace/code-block-writer/lib/index.js, imported from: file:///Users/matthew/workspace/myproject/main.ts`

However, if rather than symlinking I literally copy precisely the same directory into the same place, it seems to work.

The same error happens if I try to symlink from within `/Users/matthew/Library/Caches/deno/npm/registry.npmjs.org` to the local dev checkout.

All I want to do here is to be able to develop on a local copy of one of the dependencies for my project; I assume I'm missing something obvious. For now, I guess i'll symlink my workspace checkout to the depths of node_modules rather than vice versa as a workaround.

**Comment by dsherret** [comment]
This is now possible by using the patch feature:

```jsonc
// deno.json
{
  "patch": [
    "../path/to/local_npm_package"
  ],
  // required until Deno 2.3, but it will still be considered unstable
  "unstable": ["npm-patch"]
}
```

See overview of how it works in https://github.com/denoland/deno/pull/28512#issue-2922225164

Please try it out and provide feedback.

**Comment by leighmcculloch** [comment]
@dsherret Thanks for posting the example, it worked for me. Although, I found that I needed to also add the `nodeModulesDir` option. Is that expected? Seems like this should work without using a local/unique node modules directory.

The file I used to get it to work:
```json
{
  "imports": {
    "@stellar/stellar-base": "npm:@stellar/stellar-base@^13.1.0",
    "@stellar/stellar-sdk": "npm:@stellar/stellar-sdk@13.2.0"
  },
  "patch": [
    "../js-stellar-sdk"
  ],
  "unstable": ["npm-patch"],
  "nodeModulesDir": "auto"
}
```

This is with `deno 2.2.8`.

**Comment by brc-dd** [comment]
I'm not sure but doesn't seem to work.

Also, it's panicking if I use a tag instead of semver in imports: `"vitepress": "npm:vitepress@next"`:

<img width="1800" alt="Image" src="https://github.com/user-attachments/assets/16bef884-5054-4528-9f9b-3680a705e20c" />

**Comment by dsherret** [comment]
> Although, I found that I needed to also add the nodeModulesDir option. Is that expected?

Yes, this is required (as outlined in the issue though we should probably have a warning for this if it's not already there... haven't added documentation for this yet because it's experimental).

@brc-dd I opened https://github.com/denoland/deno/issues/28890 (I recommend opening new issues for bugs as commenting in old issues might cause it to get lost more easily)

**Comment by dsherret** [comment]
@brc-dd btw, FWIW, you can get this scenario to work by specifying `"vitepress": "npm:vitepress@2.0.0-alpha.4"` in the deno.json. I think the `*` version requirement doesn't match pre-release versions.

**Comment by brc-dd** [comment]
> I think the `*` version requirement doesn't match pre-release versions.

Ah yeah. Makes sense. For some reason, I was thinking of it more like pnpm overrides than patch. (In overrides the range doesn't matter.)
